openverse_catalog/dags/providers/provider_api_scripts/smithsonian.py

```diff
@@ -1,531 +1,443 @@
-    return source
-
-
-def _extract_tags(row, tag_types=None):
-    if tag_types is None:
-        tag_types = TAG_TYPES
-    indexed_structured = _get_indexed_structured_dict(row)
-    tag_lists_generator = (
-        _check_type(indexed_structured.get(key), list) for key in tag_types
-    )
-    return [tag for tag_list in tag_lists_generator for tag in tag_list if tag]
-
-
-def _get_descriptive_non_repeating_dict(row):
-    logger.debug("Getting descriptive_non_repeating_dict from row")
-    return _check_type(_get_content_dict(row).get("descriptiveNonRepeating"), dict)
-
-
-def _get_indexed_structured_dict(row):
-    logger.debug("Getting indexed_structured_dict from row")
-    return _check_type(_get_content_dict(row).get("indexedStructured"), dict)
+        if ordered_freetext_creator_objects:
+            first_creator_object = ordered_freetext_creator_objects[0]
+            top_priority = self.creator_types[first_creator_object["label"].lower()]
 
+            creators_list = [
+                c["content"]
+                for c in ordered_freetext_creator_objects
+                if top_priority == self.creator_types[c["label"].lower()]
+            ]
 
-def _get_freetext_dict(row):
-    logger.debug("Getting freetext_dict from row")
-    return _check_type(_get_content_dict(row).get("freetext"), dict)
-
-
-def _get_content_dict(row):
-    return _check_type(row.get("content"), dict)
-
-
-def _check_type(unknown_input, required_type):
-    """
-    This function ensures that the input is of the required type.
+            creator = (
+                "; ".join(creators_list[:-1]) + " and " + creators_list[-1]
+                if len(creators_list) > 1
+                else creators_list[0]
+            )
 
-    Required Arguments:
+        else:
+            creator = None
+
+        if creator is None:
+            logger.debug(f"No creator found in freetext:  {freetext}")
+            creator = next(indexed_structured_creator_generator, None)
+        if creator is None:
+            logger.debug(
+                f"No creator found in indexed_structured:  {indexed_structured}"
+            )
+        return creator
 
-    unknown_input:  This can be anything
-    required_type:  built-in type name/constructor
+    def _extract_tags(self, row):
+        indexed_structured = self._get_indexed_structured_dict(row)
+        tag_lists_generator = (
+            self._check_type(indexed_structured.get(key), list)
+            for key in self.tag_types
+        )
+        return [tag for tag_list in tag_lists_generator for tag in tag_list if tag]
 
-    This function will check whether the unknown_input is of type
-    required_type. If it is, it returns the unknown_input value
-    unchanged. If not, will return the DEFAULT VALUE FOR THE TYPE
-    required_type. So, if unknown_input is a float 3.0 and the
-    required_type is int, this function will return 0.
+    def ingest_records(self, **kwargs) -> None:
```

Comment:
To make sure that the unit code validation runs on the scheduled runs from Airflow, you could add it here instead of adding it in the main function:

```suggestion
    def ingest_records(self, **kwargs) -> None:
        self.validate_unit_codes_from_api()
```

Reply:
Right! I was thinking I'd have to do a sort of `preingestion_task` like in the `inatualist` script, but this solution is more straightforward. Thanks!